sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/state/DefaultSynchronousMetricStorage.java

```diff
@@ -208,25 +232,79 @@ public MetricData collect(
       aggregatorHandles = this.aggregatorHolder.aggregatorHandles;
     }
 
+    List<T> points;
+    if (memoryMode == REUSABLE_DATA) {
+      reusableResultList.clear();
+      points = reusableResultList;
+    } else {
+      points = new ArrayList<>(aggregatorHandles.size());
+    }
+
+    // In DELTA aggregation temporality each Attributes is reset to 0
+    // every time we perform a collection (by definition of DELTA).
+    // In IMMUTABLE_DATA MemoryMode, this is accomplished by removing all aggregator handles
+    // (into which the values are recorded) effectively starting from 0
+    // for each recorded Attributes.
+    // In REUSABLE_DATA MemoryMode, we strive for zero allocations. Since even removing
+    // a key-value from a map and putting it again on next recording will cost an allocation,
+    // we are keeping the aggregator handles in their map, and only reset their value once
+    // we finish collecting the aggregated value from each one.
+    // The SDK must adhere to keeping no more than maxCardinality unique Attributes in memory,
+    // hence during collect(), when the map is at full capacity, we try to clear away unused
+    // aggregator handles, so on next recording cycle using this map, there will be room for newly
+    // recorded Attributes. This comes at the expanse of memory allocations. This can be avoided
+    // if the user chooses to increase the maxCardinality.
+    if (memoryMode == REUSABLE_DATA && reset) {
+      if (aggregatorHandles.size() >= maxCardinality) {
+        aggregatorHandles.forEach(
+            (attribute, handle) -> {
+              if (!handle.isValuesRecorded()) {
+                aggregatorHandles.remove(attribute);
+              }
+            });
+      }
+    }
+
     // Grab aggregated points.
-    List<T> points = new ArrayList<>(aggregatorHandles.size());
     aggregatorHandles.forEach(
         (attributes, handle) -> {
```

Comment:
I think this can be simplified to:
```
if (!handle.isValuesRecorded()) {
  return;
}
T point = handle.aggregateThenMaybeReset(start, epochNanos, attributes, reset);;
if (reset && memoryMode == IMMUTABLE_DATA) {
  aggregatorHandlePool.offer(handle);
}
if (point != null) {
  points.add(point);
}
```

Reply:
CUMULATIVE must return data point, all the time, but in the suggestion you skip it if there are no recorded values.

Reply:
CUMULATIVE won't have a handle unless there are recorded values, so `handle.isValuesRecorded()` is always true.

Reply:
Thanks! I didn't think about it.
I simplified it per your suggestion and added javadoc explaining about memory pool per our previous comment